src/test/expression/expression_test.cpp

```diff
@@ -276,4 +276,148 @@ TEST_F(ExpressionTest, IsNullable) {
+
+  // IsNullExpression
+  expressions.emplace_back(__LINE__, is_null_(1));
+  expressions.emplace_back(__LINE__, is_null_(2));
+  expressions.emplace_back(__LINE__, is_not_null_(1));
+
+  // ListExpression
+  expressions.emplace_back(__LINE__, list_(1));
+  expressions.emplace_back(__LINE__, list_(1, 2));
+  expressions.emplace_back(__LINE__, list_(2, 3));
+
+  // LogicalExpression
+  expressions.emplace_back(__LINE__, and_(1, 0));
+  expressions.emplace_back(__LINE__, and_(1, 1));
+  expressions.emplace_back(__LINE__, or_(1, 0));
+
+  // LQPColumnExpression
+  expressions.emplace_back(__LINE__, lqp_column_(a));
+  expressions.emplace_back(__LINE__, lqp_column_(b));
+
+  // LQPSubqueryExpression
+  expressions.emplace_back(__LINE__, lqp_subquery_(int_float_node));
+  expressions.emplace_back(__LINE__, lqp_subquery_(int_float_node_nullable));
+  expressions.emplace_back(__LINE__, lqp_subquery_(int_float_node, std::make_pair(ParameterID{0}, a)));
+  expressions.emplace_back(__LINE__, lqp_subquery_(int_float_node, std::make_pair(ParameterID{1}, a)));
+  expressions.emplace_back(__LINE__, lqp_subquery_(int_float_node, std::make_pair(ParameterID{1}, b)));
+
+  // PlaceholderExpression
+  expressions.emplace_back(__LINE__, placeholder_(ParameterID{0}));
+  expressions.emplace_back(__LINE__, placeholder_(ParameterID{1}));
+
+  // PQPColumnExpression
+  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{0}, DataType::Float, false, "a + b"));
+  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{1}, DataType::Float, false, "a + b"));
+  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{1}, DataType::Int, false, "a + b"));
+  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{1}, DataType::Int, true, "a + b"));
+  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{1}, DataType::Int, true, "alias"));
+
+  // PQPSubqueryExpression
+  expressions.emplace_back(__LINE__, pqp_subquery_(std::make_shared<GetTable>("a"), DataType::Int, false));
+  expressions.emplace_back(__LINE__, pqp_subquery_(std::make_shared<GetTable>("b"), DataType::Int, false));
+  expressions.emplace_back(__LINE__, pqp_subquery_(std::make_shared<GetTable>("b"), DataType::Float, false));
+  expressions.emplace_back(__LINE__, pqp_subquery_(std::make_shared<GetTable>("b"), DataType::Float, true));
+  const auto parameter = std::make_pair(ParameterID{0}, ColumnID{0});
+  expressions.emplace_back(__LINE__, pqp_subquery_(std::make_shared<GetTable>("b"), DataType::Float, true, parameter));
+
+  // UnaryMinusExpression
+  expressions.emplace_back(__LINE__, unary_minus_(3));
+  expressions.emplace_back(__LINE__, unary_minus_(4));
+
+  // ValueExpression
+  expressions.emplace_back(__LINE__, value_(3));
+  expressions.emplace_back(__LINE__, value_(3.0f));
+  expressions.emplace_back(__LINE__, value_(3.25));
+  expressions.emplace_back(__LINE__, null_());
+
+  for (auto first_iter = expressions.begin(); first_iter != expressions.end(); ++first_iter) {
```

Comment:
I dislike loops in tests and tests that do not yield a proper stacktrace, back to the input data, on failure. So I'd prefer if each `expressions.emplace_back(__LINE__, ...` line called a helper with the assertions and instead of a all-unequal-all-others comparison test you'd test just a few explicit cases. 

Just preferences.

Reply:
Have you seen the scoped_trace? I consider that better than helper methods:


```
../src/test/expression/expression_test.cpp:408: Failure
Expected equality of these values:
  *first_expression
    Which is: SUBQUERY (PQP, 0x11169bc38)
  *deep_copy
    Which is: SUBQUERY (PQP, 0x11169c098)
Google Test trace:
../src/test/expression/expression_test.cpp:399: First expression from line 380
../src/test/expression/expression_test.cpp:408: Failure
```

Reply:
Pinging @mrzzzrm because I'm interested to hear if your loop aversion includes loops that have those traces

Reply:
Yes, saw that. Makes is it a bit better, but not good. 

A failing test should be as easy as possible to understand by even a non-author. The best the test-author can do for this, imo, is make sure that all code/data involved in the test is in the stacktrace when the failing test is run in the debugger. If we make sure that's the case then any test is follows the same pattern and is "familiar" even to non-authors. If every test follows it's own pattern and requires the non-author to parse the log for line X and Y, then to parse the entire test function (which might well be the thing that's broken!), this is a significant burden. 
(There is a fascinating extent to which people complicate their tests that makes it very hard to figure out what is actually broken (test or code?) and what is actually being tested in a test case. I'm just trying to avoid stepping into that territory, here. Keeping things simple paranoidly is imo just good enough.)

Reply:
But please take this as a philosophical rant about how unit tests should be written imho.

Reply:
I agree with the general goal of making test failures easy to understand. At the same time, it should be easy to check that tests are complete (i.e., cover each variation in the input data). For me, it is easy to see that this list covers all EXIST cases:

```
  expressions.emplace_back(__LINE__, exists_(lqp_subquery_(int_float_node)));
  expressions.emplace_back(__LINE__, not_exists_(lqp_subquery_(int_float_node)));
  expressions.emplace_back(__LINE__, not_exists_(lqp_subquery_(int_float_node_nullable)));
```

A test that does the same without loops could look like this:

```
TEST_F(ExpressionTest, EqualsAndHash2) {
  const auto test_reflection = [](const auto& expression) {
    const auto deep_copy = expression->deep_copy();
    EXPECT_EQ(*expression, *deep_copy);
    EXPECT_EQ(expression->hash(), deep_copy->hash());
  };

  test_reflection(exists_(lqp_subquery_(int_float_node)));
  test_reflection(not_exists_(lqp_subquery_(int_float_node)));
  test_reflection(not_exists_(lqp_subquery_(int_float_node_nullable)));

  EXPECT_NE(*exists_(lqp_subquery_(int_float_node)), *not_exists_(lqp_subquery_(int_float_node)));
  EXPECT_NE(*exists_(lqp_subquery_(int_float_node)), *not_exists_(lqp_subquery_(int_float_node_nullable)));
}
```

To avoid adding the expressions twice, I could build a helper that first checks the reflective equality and the hash of the first argument and then compares the first with the second argument:

```
TEST_F(ExpressionTest, EqualsAndHash2) {
  const auto test = [](const auto& left, const auto& right) {
    const auto deep_copy = left->deep_copy();
    EXPECT_EQ(*left, *deep_copy);
    EXPECT_EQ(left->hash(), deep_copy->hash());

    EXPECT_NE(*left, *right);
  };

  test(*exists_(lqp_subquery_(int_float_node)), *not_exists_(lqp_subquery_(int_float_node)));
  test(*exists_(lqp_subquery_(int_float_node)), *not_exists_(lqp_subquery_(int_float_node_nullable)));
}
```

This makes it hard to see that `not_exists_` is never checked against its deep copy. Also, if I undo the fix for EqualsExpression, the error message looks like this:

```
[ RUN      ] ExpressionTest.EqualsAndHash2
../src/test/expression/expression_test.cpp:430: Failure
Expected: (*left) != (*right), actual: 64-byte object <D0-60 73-04 01-00 00-00 18-83 8E-0C 01-00 00-00 00-83 8E-0C 01-00 00-00 07-00 00-00 4E-75 6C-6C 80-01 BE-0C 01-00 00-00 90-01 BE-0C 01-00 00-00 90-01 BE-0C 01-00 00-00 00-00 00-00 69-67 6E-65> vs 64-byte object <D0-60 73-04 01-00 00-00 D8-D6 8C-0C 01-00 00-00 C0-D6 8C-0C 01-00 00-00 07-00 00-00 6E-6B 4F-66 90-01 BE-0C 01-00 00-00 A0-01 BE-0C 01-00 00-00 A0-01 BE-0C 01-00 00-00 01-00 00-00 69-67 6E-65>
```

That's hardly better. Even if we got gtest to print the column name, it would not be perfect, as it does not give us the absolute information on how the expression was created.

So while I agree with you that generated data and loops make it hard for the reader to understand what's going wrong (OperatorsIndexScanTest.SingleColumnScanOnDataTable, only to name a randomly picked one), I wouldn't consider it a black and white issue. In this case, my tendency is to stick to the version in the PR.

Reply:
> I wouldn't consider it a black and white issue. In this case, my tendency is to stick to the version in the PR.

Yes, and alright.

Reply:
(But it might be worth trying to figure out why googletest doesn't pretty-print these expressions so you could get rid of the weird capturing of __LINE__ there...)

Reply:
> But it might be worth trying to figure out why googletest doesn't pretty-print these expressions

That is because `auto` in the test is not an AbstractExpression but an ExistsExpression. No idea why googletest doesn't use `operator<<` of the parent class while a normal print does.

> so you could get rid of the weird capturing of `__LINE__` there...

I'd still leave `__LINE__` in, as `as_column_name` is incomplete:

```
../src/test/expression/expression_test.cpp:415: Failure
Expected equality of these values:
  *first_expression
    Which is: a + b
  *second_expression
    Which is: a + b
Google Test trace:
../src/test/expression/expression_test.cpp:414: Second expression from line 374
../src/test/expression/expression_test.cpp:399: First expression from line 373
../src/test/expression/expression_test.cpp:415: Failure
```

for 

```
  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{0}, DataType::Float, false, "a + b"));
  expressions.emplace_back(__LINE__, pqp_column_(ColumnID{1}, DataType::Float, false, "a + b"));
```